<div align="center">
  <h3><b>ESCUELA POLITÉCNICA NACIONAL</b></h3>
  <h3><b>FACULTAD DE INGENIERÍA EN SISTEMAS</b></h3>
  <h3><b>INGENIERÍA EN CIENCIAS DE LA COMPUTACIÓN</b></h3>
  <h3><b>RECUPERACIÓN DE LA INFORMACIÓN</b></h3>
</div>

---
**Nombre**   Mark Hernández        
**Fecha**    04/05/26  
**Docente**  Iván Carrera

# Ejercicio 3: Modelo Vectorial y TF-IDF

## Objetivo de la práctica

- Comprender el modelo vectorial como base para representar documentos y consultas.
- Calcular la matriz TF-IDF para el corpus `data/01_corpus_turismo_500.txt`
- Calcular la matriz TF-IDF para el corpus `Gutenberg 1000`

### Paso 1: Calcular la matriz TF-IDF para el corpus `data/01_corpus_turismo_500.txt`

Primero obtenemos el corpus del archivo .txt

In [1]:
import os
import re

path = "..\\03tfidf\\data"
file = os.listdir(path)

contenido = []

with open(os.path.join(path, file[0]),"r", encoding="utf-8") as f:
    contenido = f.read().lower()
    palabras = re.findall(r'\w+', contenido)
    vocabulario = set(palabras)

print(list(vocabulario)[:100])

print(len(vocabulario))


['invita', 'única', 'se', 'como', 'los', 'mercado', 'destaca', 'lagunas', 'senderismo', 'muchos', 'tranquilo', 'spondylus', 'una', 'a', 'disfrutan', 'sorprende', 'deportes', 'deslumbra', 'cajas', 'entre', 'colonial', 'país', 'amazonía', 'de', 'popular', 'cotopaxi', 'canopy', 'quilotoa', 'turquesa', 'caminatas', 'experiencia', 'parque', 'laguna', 'es', 'visitar', 'típica', 'aventura', 'las', 'próximo', 'y', 'vilcabamba', 'interesados', 'destino', '2000', 'patrimonio', 'ecológico', 'conecta', 'quito', 'islas', 'ruta', 'llena', 'la', 'fiestas', 'surf', 'humanidad', 'un', 'conocido', 'turismo', 'playas', 'guayaquil', 'arquitectura', 'mindo', 'otavalo', 'para', 'sorprendente', 'agua', 'ecuatoriana', 'nacional', 'su', 'biodiversidad', 'rafting', 'megadiverso', 'indígena', 'por', 'montañita', 'malecón', 'lugar', 'vida', 'longevidad', 'naturaleza', 'galápagos', 'ofrece', 'local', 'volcán', 'cuenca', 'ideal', 'atraen', 'inolvidable', 'perfecto', 'baños', 'feriado', 'famoso', 'atrae', 'tiene', '

Luego procedemos a calcular los valores de frecuencia con que aparece cada término dentro del documento. Para ello hacemos uso del concepto de índice invertido.

In [2]:
import pandas as pd

indice_invertido = {}

for palabra in vocabulario:
    indice_invertido[palabra] = 0

for palabra in re.findall(r'\w+', contenido): 
    if palabra in indice_invertido.keys():
        indice_invertido[palabra] += 1

print(len(indice_invertido))
    
t1 = pd.DataFrame(list(indice_invertido.items()), columns=["palabra", "frecuencia"])
t1.head(20)

118


,palabra,frecuencia
0,invita,37
1,única,33
2,se,27
3,como,79
4,los,59
5,mercado,33
6,destaca,32
7,lagunas,37
8,senderismo,44
9,muchos,27


Con estas frecuencias calculamos el valor de tf.

In [3]:
import math
tf = {}
for palabra, frec in indice_invertido.items():
    tf[palabra] = frec/len(palabras)

Tabla_tf = pd.DataFrame(list(tf.items()), columns=["palabra", "tf"])
Tabla_tf.head(20)

,palabra,tf
0,invita,0.006055
1,única,0.005400
2,se,0.004418
3,como,0.012928
4,los,0.009655
5,mercado,0.005400
6,destaca,0.005236
7,lagunas,0.006055
8,senderismo,0.007200
9,muchos,0.004418


Con el valor de tf ya calculado, procedemos a calcular el valor de idf. Para este primer caso, hay que considerar que al ser un único documento, N=1 y la frecuencia con que aparece un término dentro de este documento es del 100%, es decir, df_t = 1. Así, el valor de idf es el mismo para todos debido a que estamos evaluando un solo libro, lo que nos indica que los términos del corpus no son muy informativos al compararlos en un solo documento.

In [4]:
N = len(file)
df_t=1 # Todas las palabras aparecen en el 100% de los documentos, ya que nuestra colección es solo el corpus_500.
idf = {}
for palabra, frec in indice_invertido.items():
    idf[palabra] = math.log10(N/df_t)
    
Tabla_idf = pd.DataFrame(list(idf.items()), columns=["palabra", "idf"])
Tabla_idf.head(20)    

,palabra,idf
0,invita,0.0
1,única,0.0
2,se,0.0
3,como,0.0
4,los,0.0
5,mercado,0.0
6,destaca,0.0
7,lagunas,0.0
8,senderismo,0.0
9,muchos,0.0


Puesto que estamos usando un único documento y la matriz TF_IDF evalúa la importancia de una palabra de un documento en relación a una colección de documentos, los elementos de dicha matriz serán equivalentes a contar la frecuencia de las palabras y normalizarlas

In [5]:
TF_IDF = {}
norma = math.sqrt(sum([tf[palabra]**2 for palabra in indice_invertido.keys()]))
for palabra in indice_invertido.keys():
    TF_IDF[palabra] = tf[palabra] /norma

Tabla_TF_IDF = pd.DataFrame([TF_IDF])
Tabla_TF_IDF.head(20)

,invita,única,se,como,los,mercado,destaca,lagunas,senderismo,muchos,...,famosas,histórico,nocturna,artesanía,en,visitan,santa,del,gastronomía,espectacular
0,0.047375,0.042254,0.034571,0.101153,0.075544,0.042254,0.040973,0.047375,0.056338,0.034571,...,0.060179,0.025608,0.042254,0.042254,0.034571,0.034571,0.040973,0.060179,0.084507,0.025608


Realizando con la librería sklearn

In [6]:
from sklearn.feature_extraction.text import TfidfVectorizer

documento = [contenido]
# Inicializar el vectorizador
vectorizer_individual = TfidfVectorizer()

# Ajustar el modelo y transformar el texto en la matriz TF-IDF
matriz_tfidf = vectorizer_individual.fit_transform(documento)

# Visualizar los resultados usando Pandas
nombres_caracteristicas = vectorizer_individual.get_feature_names_out()
df_tfidf = pd.DataFrame(matriz_tfidf.toarray(), columns=nombres_caracteristicas)

print("Matriz TF-IDF para un solo documento:\n")
df_tfidf.head(20)

Matriz TF-IDF para un solo documento:



,2000,agua,amazonía,arquitectura,artesanía,atrae,atraen,auténtico,aventura,aves,...,típica,un,una,vida,vilcabamba,visitan,visitantes,visitar,volcán,única
0,0.043208,0.041899,0.035352,0.044517,0.043208,0.035352,0.043208,0.017021,0.041899,0.057611,...,0.041899,0.229134,0.125696,0.043208,0.035352,0.035352,0.07856,0.120459,0.035352,0.043208


Es fácil observar que los valores son bastante cercanos. La diferencia en sus decimales puede deberse a que no se limpiaron las palabras de la manera correcta en la manera manual, mientras que sklearn lo realizó un poco mejor. Se puede observar que el primero tiene 2 palabras más que al usar sklearn (118 vs 116)

### Paso 2: Construir el corpus `Gutenberg 1000`

El corpus `Gutenberg 1000` es un corpus compuesto por 1000 libros de Gutenberg Project

In [ ]:
import re

path2 = "..\\Data\\Gutenberg_1000"
files = os.listdir(path2)
contenidos_libros = []
nombres_libros = []
indice_invertido = {}
for file in files:
    # Se iteran los archivos
    with open(path2 + "\\" + file, "r", encoding="utf-8") as f:
        contenido = f.read().lower()
        # Extrae solo palabras (letras y números)
        palabras = re.findall(r'\w+', contenido)
        contenidos_libros.append(contenido)
        nombres_libros.append(file)
        vocabulario = set(palabras)
        for palabra in vocabulario:
            if palabra in indice_invertido:
                indice_invertido[palabra].append(file)
            else:
                indice_invertido[palabra] = [file]
print('Número de palabras del índice invertido: ',len(indice_invertido))
print(f"✓ Libros cargados: {len(contenidos_libros)}")


Número de palabras del índice invertido:  1061431
✓ Libros cargados: 1000


In [8]:
Tabla_archivos = pd.DataFrame(list(indice_invertido.items()), columns=['Palabra', 'Archivos'])
Tabla_archivos.head(20)

,Palabra,Archivos
0,illuminate,"[Book1.txt, book10.txt, book10008.txt, book105..."
1,gaze,"[Book1.txt, book10.txt, book100.txt, book10008..."
2,approve,"[Book1.txt, book10.txt, book100.txt, book10392..."
3,trace,"[Book1.txt, book10.txt, book100.txt, book10008..."
4,mercy,"[Book1.txt, book10.txt, book100.txt, book10008..."
5,renewing,"[Book1.txt, book100.txt, book10372.txt, book11..."
6,heavens,"[Book1.txt, book10.txt, book100.txt, book10008..."
7,breathed,"[Book1.txt, book10.txt, book10008.txt, book100..."
8,whole,"[Book1.txt, book10.txt, book100.txt, book10008..."
9,internal,"[Book1.txt, book10.txt, book100.txt, book10008..."


### Paso 3: Calcular la matriz TF-IDF para el corpus `Gutenberg 1000`

Primero importamos las librerías necesarias.

In [9]:
import time
from sklearn.feature_extraction.text import TfidfVectorizer

Con el contenido pre-cargado realizamos el cálculo de la mmatriz TF-IDF.

In [10]:
# Crear el vectorizador TF-IDF para Gutenberg 1000
# - sublinear_tf=True: TF logarítmico (1 + log(tf)) para amortiguar términos muy frecuentes
# - min_df=2:          ignorar términos que aparecen en menos de 2 libros
# - max_df=0.85:       ignorar términos en más del 85% de los libros (palabras muy comunes)
# - max_features=50000: limitar el vocabulario a los 50.000 términos más frecuentes
# - stop_words='english': eliminar stopwords en inglés
vectorizador_gutenberg = TfidfVectorizer(
    sublinear_tf=True,
    min_df=2,
    max_df=0.85,
    max_features=50000,
    stop_words="english",
    encoding="utf-8",
    decode_error="replace"
)

start_time = time.time()
print("Calculando matriz TF-IDF para Gutenberg 1000...")

# fit_transform aprende el vocabulario y construye la matriz (libros x términos)
matriz_tfidf_gutenberg = vectorizador_gutenberg.fit_transform(contenidos_libros)

elapsed = time.time() - start_time
print(f"✓ Matriz calculada en {elapsed:.2f} segundos")
print(f"\nDimensiones de la matriz TF-IDF (Gutenberg 1000):")
print(f"  → {matriz_tfidf_gutenberg.shape[0]} documentos  x  {matriz_tfidf_gutenberg.shape[1]} términos únicos")
print(f"\nTamaño en memoria (sparse): {matriz_tfidf_gutenberg.data.nbytes / 1024 / 1024:.2f} MB")
print(f"Elementos no cero: {matriz_tfidf_gutenberg.nnz:,}")

Calculando matriz TF-IDF para Gutenberg 1000...
✓ Matriz calculada en 128.26 segundos

Dimensiones de la matriz TF-IDF (Gutenberg 1000):
  → 1000 documentos  x  50000 términos únicos

Tamaño en memoria (sparse): 37.29 MB
Elementos no cero: 4,887,463


In [11]:
# Visualizar estadísticas de la matriz
terminos_gutenberg = vectorizador_gutenberg.get_feature_names_out()

# Top 20 términos con mayor IDF (los más raros / específicos)
import numpy as np
idf_values = vectorizador_gutenberg.idf_
top_idf_idx = np.argsort(idf_values)[::-1][:20]

print("Top 20 términos con mayor IDF (más específicos del corpus):")
for idx in top_idf_idx:
    print(f"  {terminos_gutenberg[idx]:<25} IDF = {idf_values[idx]:.4f}")

# Top 20 términos con menor IDF (los más comunes)
bot_idf_idx = np.argsort(idf_values)[:20]
print("\nTop 20 términos con menor IDF (más comunes):")
for idx in bot_idf_idx:
    print(f"  {terminos_gutenberg[idx]:<25} IDF = {idf_values[idx]:.4f}")

Top 20 términos con mayor IDF (más específicos del corpus):
  一卷                        IDF = 6.8101
  bolsheviki                IDF = 6.8101
  bondel                    IDF = 6.8101
  ripoll                    IDF = 6.8101
  ών                        IDF = 6.8101
  σωκράτη                   IDF = 6.8101
  υπέρ                      IDF = 6.8101
  morisque                  IDF = 6.8101
  blondine                  IDF = 6.8101
  bluegum                   IDF = 6.8101
  bracknell                 IDF = 6.8101
  bowne                     IDF = 6.8101
  potassii                  IDF = 6.8101
  bourienne                 IDF = 6.8101
  porthos                   IDF = 6.8101
  bruoder                   IDF = 6.8101
  brownlows                 IDF = 6.8101
  bruck                     IDF = 6.8101
  broadley                  IDF = 6.8101
  brocklehurst              IDF = 6.8101

Top 20 términos con menor IDF (más comunes):
  years                     IDF = 1.1623
  know                      IDF =

### Paso 4: Programar una función `buscar()` para el corpus `Gutenberg 1000`

In [12]:
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

def buscar(consulta: str, top_n: int = 10) -> pd.DataFrame:
    """
    Busca los documentos más relevantes para una consulta usando el modelo vectorial TF-IDF.

    El proceso es:
      1. Transformar la consulta al espacio vectorial TF-IDF con el mismo vocabulario
         que fue aprendido al construir la matriz (fit_transform del corpus).
      2. Calcular la similitud coseno entre el vector de la consulta y todos los documentos.
      3. Retornar los top_n documentos ordenados por similitud descendente.

    Parámetros:
        consulta (str): Texto de búsqueda libre.
        top_n   (int): Número de resultados a retornar (por defecto 10).

    Retorna:
        pd.DataFrame con columnas [rank, libro, similitud]
    """
    start = time.time()

    # Transformar la consulta usando el vectorizador ya ajustado (solo transform, no fit)
    vector_consulta = vectorizador_gutenberg.transform([consulta])

    # Calcular similitud coseno entre la consulta y todos los documentos
    # cosine_similarity retorna un array de forma (1, n_documentos)
    similitudes = cosine_similarity(vector_consulta, matriz_tfidf_gutenberg).flatten()

    # Obtener los índices de los top_n documentos más similares
    top_indices = np.argsort(similitudes)[::-1][:top_n]

    elapsed = time.time() - start

    # Construir el DataFrame de resultados
    resultados = pd.DataFrame({
        "rank":       range(1, len(top_indices) + 1),
        "libro":      [nombres_libros[i] for i in top_indices],
        "similitud":  [round(similitudes[i], 6) for i in top_indices]
    })

    print(f"Consulta: '{consulta}'")
    print(f"Tiempo de búsqueda: {elapsed:.4f} segundos")
    print(f"Documentos con similitud > 0: {np.sum(similitudes > 0)}")
    return resultados

### Prueba 1: búsqueda de un tema específico
----

In [13]:
resultados = buscar("whale sea adventure ocean voyage", top_n=10)
print(resultados.to_string(index=False))

Consulta: 'whale sea adventure ocean voyage'
Tiempo de búsqueda: 0.3326 segundos
Documentos con similitud > 0: 789
 rank         libro  similitud
    1 book11105.txt   0.120206
    2 book57091.txt   0.059361
    3 book41221.txt   0.052793
    4 book44413.txt   0.052046
    5 book34392.txt   0.048929
    6 book52462.txt   0.047420
    7 book10339.txt   0.047359
    8 book13442.txt   0.043886
    9     Book2.txt   0.043418
   10 book55479.txt   0.041350


### Prueba 2: búsqueda de amor y romance
----

In [14]:
resultados2 = buscar("love marriage romance society", top_n=10)
print(resultados2.to_string(index=False))

Consulta: 'love marriage romance society'
Tiempo de búsqueda: 0.3094 segundos
Documentos con similitud > 0: 786
 rank         libro  similitud
    1 book47697.txt   0.048561
    2 book11970.txt   0.042120
    3    book17.txt   0.041637
    4 book51945.txt   0.040890
    5 book14679.txt   0.040551
    6 book12038.txt   0.040169
    7    book26.txt   0.039984
    8 book40006.txt   0.039486
    9 book12563.txt   0.038783
   10 book11235.txt   0.038637


### Prueba 3: búsqueda de ciencia y tecnología
----

In [15]:
resultados3 = buscar("science experiment laboratory discovery", top_n=10)
print(resultados3.to_string(index=False))

Consulta: 'science experiment laboratory discovery'
Tiempo de búsqueda: 0.4358 segundos
Documentos con similitud > 0: 713
 rank         libro  similitud
    1 book42626.txt   0.081721
    2 book51231.txt   0.075974
    3 book49377.txt   0.057133
    4    book11.txt   0.056561
    5 book50783.txt   0.055242
    6 book41375.txt   0.046373
    7 book59706.txt   0.044531
    8  book3990.txt   0.042636
    9 book54609.txt   0.042319
   10 book10008.txt   0.041702
